cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin.yaml

```diff
@@ -114,3 +115,7 @@ spec:
         - name: lib-modules
           hostPath:
             path: /lib/modules
+        - name: ceph-csi-config
```

Comment:
What is the layout of the CSI pods? We have a single provisioner, a single attacher, and a daemonset for the driver on each node? 

Can we run multiple provisioners and attachers (one for each rook-ceph cluster)? If so, that means the provisioner and attacher would be running in the same namespace as the rook-ceph cluster and reading the configmap will be obvious in the same namespace. If not, I assume the provisioner and attacher need to run in the same namespace as the operator and the same configmap will contain the mon endpoints for each cluster being managed by the operator.

Reply:
Multiple provisioners (one for each rook-ceph cluster) can be run as long as [this](https://github.com/rook/rook/blob/7d3878378d7bf8a29d4c902bb07401968e3b0428/cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner.yaml#L22) driver address value is made unique per instance.

A single instance can also be run, but in that case, as noted by you, the configmap will contain mon end points for each cluster being managed by the operator.

Reply:
Looks like this approach will only work for a single namespace. According to the [configmap restrictions](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#restrictions):
```
A ConfigMap can only be referenced by pods residing in the same namespace.
```
So the csi driver won't be able to mount the configmap unless it is stored in the same namespace. To support that scenario, the operator would have to aggregate them into a single configmap. What is the `ADDRESS` you referred to? Can we assume it will be unique and we will run the CSI driver per ceph cluster?

Reply:
> So the csi driver won't be able to mount the configmap unless it is stored in the same namespace. To support that scenario, the operator would have to aggregate them into a single configmap. What is the `ADDRESS` you referred to? Can we assume it will be unique and we will run the CSI driver per ceph cluster?

The ADDRESS is the CSI driver name, which should be unique per CSI driver/instance. The usual names for Ceph-CSI are, `rbd.csi.ceph.com` and `cephfs.csi.ceph.com`. These are further referenced in the StorageClass using the `provisioner` key to direct requests to the right CSI driver and its instance.

For Rook CSI instance per cluster can specialize this name as `clusterNamespace`+`rbd/cephfs.csi.ceph.com` and hence retain the configmap with the same namespace as the Rook cluster, where the CSI pods would also be running. This also enables running distinct per cluster Ceph-CSI instances.